### DISCUSSION TITLE
Best Query Code Management

### DISCUSSION THREAD

**Discussion by liverpool1026** [sqlalchemy/sqlalchemy#8463] [Usage Questions] [answered]
Hi there,

Just wondering what's the suggested way of managing query code within a project.

e.g.
I have a Table A
| id | col A | col B | col C |
|----|-------|-------|-------|

and a Table B
| id | col D | col E |
|----|-------|-------|

Table B's col D fk to Table A.

If I have two use case,

1. I need a query which joins A and B and need col A and col E
2. I need a query which joins A and B and need col A and col B and col E

What's the preferred way?

1. Write two seperate queries at the call-site
`session.query(TableA.colA, TableB.colE).join(TableB, TableA.id == TableB.colD)`
and
`session.query(TableA.colA, TableA.colB, TableB.colE).join(TableB, TableA.id == TableB.colD)`

2. Reuse the query that contains all the data
`session.query(TableA.colA, TableA.colB, TableB.colE).join(TableB, TableA.id == TableB.colD)`
So both call-site uses this query, but for the case that doesn't need colB, it will just get extra data

3. Make the most simple query and modify it later
`session.query(TableA.colA, TableB.colE).join(TableB, TableA.id == TableB.colD)`
And on the one that needs colB, we do
`query = query.add_columns(TableA.colB)`

4. Some other ways 

This is just a simple example, the actual queries will be much more complicated. Just seeking some opinions.

Thanks

**Comment by zzzeek** [comment, marked as answer]
I would have Python functions that create the queries you want each time.     if desirable, the functions could call down to one main function that chooses the desired query to create based on parameters passed to the function.

In particular, the legacy "session.query()" construct is a poor choice for "reuse" because it's linked to a specific `Session`.   The `Query` can be assigned to a different `Session` using `Query.with_session(other_session)` but overall Query isn't really meant to be hanging around for "re-use" later on.

**Reply by CaselIT** [reply]
select() for this is a lot better, since you can construct them one and re-use multiple times

**Reply by liverpool1026** [reply]
Thanks for the response, appreciated that.

Session is probably not the issue here for us, we will have something like
```
def get_query(session) -> sa.orm.Query:
       return session.query(...)
```
@zzzeek
Is you suggestion to make function like:
```
def get_query(session, cols) -> sa.orm.Query:
      return session.query(*cols).join(...)
```

@CaselIT 
Are you saying the new way of querying is the better to support reusing? We are still on 1.3 so we are still stuck with the old style of querying and I don't know much about the new style of query using select(). Can you provide me a simple example.

Thanks again.

**Reply by CaselIT** [reply]
Here is a simple example; https://docs.sqlalchemy.org/en/14/orm/quickstart.html#simple-select
select does not need a session on creation, only when executing it.

But since you are using 1.3 that syntax will not work (it will return core rows I think).

As an alternative to select, you may want to use `Query` directly, so that you can create a query direct;y
```py
def get_query(cols) -> sa.orm.Query:
      return sa.orm.Query(*cols).join(...)

**Reply by liverpool1026** [reply]
@CaselIT Thanks for the quick response. I now see what you mean by the new way is better to manage (because session is not required at construction).

I think my question is more around how to dynamically create queries for different use case.
```
def get_query(cols) -> sa.orm.Query:
      return sa.orm.Query(*cols).join(...)
```
Can handle dynamic columns, but it doesn't guard against passing in columns on different tables (which may require join).

@zzzeek mentioned but overall Query isn't really meant to be hanging around for "re-use" later on.

So I assume the recommended practice from sqlalchemy should be explicitly writing query at each call-site instead of finding a way to dynamically create and reuse it?

**Reply by CaselIT** [reply]
> @zzzeek mentioned but overall Query isn't really meant to be hanging around for "re-use" later on

I think what he meant is that an instance of query is used only one to query. If you use a function as factory than you should be ok, since every time you would get a new query instance that you can use

**Reply by liverpool1026** [reply]
@CaselIT  I see.

One last question, do you personally have or know any examples of having a factory function to generate queries?

Thanks again.

**Reply by CaselIT** [reply]
sure, I've used in the past at work.

the new style select are used in multiple points in the codebase. Just some examples
https://github.com/sqlalchemy/sqlalchemy/blob/d3e0b8e750d864766148cdf1a658a601079eed46/lib/sqlalchemy/dialects/oracle/base.py#L1611-L1631
https://github.com/sqlalchemy/sqlalchemy/blob/d3e0b8e750d864766148cdf1a658a601079eed46/lib/sqlalchemy/dialects/oracle/base.py#L2040-L2083
https://github.com/sqlalchemy/sqlalchemy/blob/d3e0b8e750d864766148cdf1a658a601079eed46/lib/sqlalchemy/dialects/postgresql/base.py#L4251-L4274
